openedx_authz/api/data.py

```diff
@@ -0,0 +1,227 @@
+"""Data classes and enums for representing roles, permissions, and policies."""
+
+from enum import Enum
+from typing import Literal, Type
+
+from attrs import define
+
+
+class GroupingPolicyIndex(Enum):
+    """Index of fields in a grouping policy."""
+
+    SUBJECT = 0
+    ROLE = 1
+    SCOPE = 2
+    # The rest of the fields are optional and can be ignored for now
+
+
+class PolicyIndex(Enum):
+    """Index of fields in a policy."""
+
+    ROLE = 0
+    ACT = 1
+    SCOPE = 2
+    EFFECT = 3
+    # The rest of the fields are optional and can be ignored for now
+
+
+@define
+class AuthZData:
+    """Base class for all authz data classes.
+
+    Attributes:
+        NAMESPACE: The namespace prefix for the data type (e.g., 'user', 'role').
+        SEPARATOR: The separator between the namespace and the identifier (e.g., ':', '@').
+        external_key: The ID for the object outside of the authz system (e.g., username).
+            Could also be used for human-readable names (e.g., role or action name).
+        namespaced_key: The ID for the object within the authz system (e.g., 'user@john_doe').
+    """
+
+    SEPARATOR: str = "@"
+    NAMESPACE: str = None
+
+    external_key: str = ""
+    namespaced_key: str = ""
+
+    def __attrs_post_init__(self):
+        """Post-initialization processing for attributes.
+
+        This method ensures that either external_key or namespaced_key is provided,
+        and derives the other attribute based on the NAMESPACE and SEPARATOR.
+
+        Note:
+        I will always instantiate with either external_key or namespaced_key, never both.
+        So we need to derive the other one based on the NAMESPACE.
+        """
+        if self.NAMESPACE and not self.namespaced_key:
+            self.namespaced_key = f"{self.NAMESPACE}{self.SEPARATOR}{self.external_key}"
```

Comment:
Should we throw errors in cases where `self.NAMESPACE` is None/empty, or both / neither `external_key` & `namespaced_key` are set? Or is it being validated in some other way that one or the other is non-empty?